### PR TITLE
Use icu shards instead of loading icudt.dat

### PR DESCRIPTION
**Pull request by pranavkm**
### Description

.NET runtime started shipping globalization data in shards. Loading the shard based on the browser culture reduces the WASM download size which is desirable.

### Customer impact

Reduced download size

### Risk

Medium. Apps that configure the culture in .NET would require opting out of this feature. We attempt to detect this case, and we have tests for it, but we haven't seen how it behaves in the wild.

-----------------------------------

This change defaults Blazor to load icu shards rather than the full icu data dropping the size to 2046.15 KB from the current 2220.16 KB (~160KB reduction in size).

One of the issues that is being investigated is that loading icu data needs to happen using ambient data before the runtime has been loaded and not the .NET specified culture. We use navigator.languages which is the same property that is used by Emscripten to set the default culture. In addition, I've added a JS API that will allow users to specify a culture as part of initializing Blazor WebAssembly. This may help in the niche cases where the culture is readily known but users do not want to pay the cost of loading the full icu data.

I did not want to rely on the new API for our tests until we were happy with it. At the same time configuring browsers to launch with a specific culture requires all other browser instances to be gone. So writing a test that exercises the non-EFIGS data is a bit challenging.

### CODE REVIEW

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 1]
```diff
@@ -138,13 +138,13 @@ export const monoPlatform: Platform = {
     return ((baseAddress as any as number) + (fieldOffset || 0)) as any as T;
   },
 
-  beginHeapLock: function() {
+  beginHeapLock: function () {
```
Comment (pranavkm) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:141]: Sorry about this - I formatted the file and it made a bunch of changes. Use the Hide whitespaces option to avoid seeing unnecessary changes.

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 2]
```diff
@@ -488,16 +505,31 @@ async function loadTimezone(timeZoneResource: LoadingResource) : Promise<void> {
...
 }
 
-async function loadICUData(icuDataResource: LoadingResource) : Promise<void> {
+function getICUResourceName(bootConfig: BootJsonData, culture: string | undefined): string {
+  const combinedICUResourceName = 'icudt.dat';
+  if (!culture || bootConfig.icuDataMode == ICUDataMode.All) {
+    return combinedICUResourceName;
+  }
+
+  const prefix = culture.split('-')[0];
```
Comment (pranavkm) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:514]: @lewing

File: src/Components/test/E2ETest/Tests/WebAssemblyGlobalizationTest.cs [hunk 3]
```diff
@@ -14,9 +14,7 @@ namespace Microsoft.AspNetCore.Components.E2ETest.Tests
 {
     // For now this is limited to server-side execution because we don't have the ability to set the
     // culture in client-side Blazor.
-    // This type is internal since localization currently does not work.
-    // Make it public onc https://github.com/dotnet/runtime/issues/38124 is resolved.
-    internal class WebAssemblyGlobalizationTest : GlobalizationTest<ToggleExecutionModeServerFixture<Program>>
```
Comment (pranavkm) [inline src/Components/test/E2ETest/Tests/WebAssemblyGlobalizationTest.cs:19]: Oops

File: src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            var localizedDisplay = Browser.Exists(By.Id("localizedString"));
+            Assert.Equal("Hello", localizedDisplay.Text); // No localized resources for this culture.
+        }
+
+        [Fact(Skip = "Figure out why this is broken")]
+        public void LoadingApp_KannadaLanguage_Works()
+        {
+            // Arrange
+            // This verifies the non-CJK icu data set.
+            var culture = new CultureInfo("kn");
```
Comment (pranavkm) [inline src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs:90]: @lewing \ @EgorBo I authored a few tests to verify that Blazor is loading icu resources correctly one for each of the 3 shards. I verified that we're loading the right shard (and there's another test that verifies this for the `ru` culture), but this one appears to be failing. 

```
    Assert.Equal() Failure
               ↓ (pos 1)
    Expected: 2/9/2020 12:00:00 ಪೂರ್ವಾಹ್ನ
    Actual:   2020-09-02 00:00:00
```

File: src/Components/Web.JS/src/Platform/BootConfig.ts [hunk 5]
```diff
@@ -37,3 +38,9 @@ export interface ResourceGroups {
 }
 
 export type ResourceList = { [name: string]: string };
+
+export enum ICUDataMode {
+  Default,
+  All,
+  Invariant
+}
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Platform/BootConfig.ts:46]: On naming, do you think `Sharded` might be clearer than `Default`? Sometimes "default" means "one of the other values but I'm not specifying which" whereas here it's a different behavior than either of the other two.

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:141]: No problem. Best not to fight the tools :) If this is how VS code wants to format it, let's do it like that!

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 6]
```diff
@@ -302,16 +304,27 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
...
+      const applicationCulture = resourceLoader.startOptions.applicationCulture || (navigator.languages && navigator.languages[0]);
+
+      if (resourceLoader.bootConfig.icuDataMode == ICUDataMode.Default && culturesToLoad && culturesToLoad[0] !== applicationCulture) {
+        // We load an initial icu file based on the browser's locale. However if the application's culture requires a different set, flag this as an error.
+        if (resourceLoader.bootConfig.debugBuild) {
+          throw new Error('This application\'s globalization settings requires using the combined globalization data file. ' +
+          'To configure your application to use this data file, set <BlazorWebAssemblyLoadAllGlobalizationData>true</BlazorWebAssemblyLoadAllGlobalizationData> in the application\'s project file. ');
+        } else {
+          throw new Error('Globalization data files required to correctly execute this application are missing.');
+        }
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:319]: Not sure it's useful to distinguish debug/prod cases here, since in both cases only the application developer will see this or do anything about it. Also it would be good to clarify *why* the situation has occurred. Suggestion:

```suggestion
        throw new Error('To change culture dynamically during startup, set <BlazorWebAssemblyLoadAllGlobalizationData>true</BlazorWebAssemblyLoadAllGlobalizationData> in the application\'s project file.');
```

Is it true that this only happens due to dynamic change of thread culture? Could it ever happen because, for example, the browser says it's in a culture that .NET doesn't even know about the existence of? If the latter could ever happen that's a bit of a bearpit - hope it's not an error by default!

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:510]: Yeah, it would be really good if this logic was inside the runtime, for example as a function like `MONO.getShardFilenameForCulture(cultureName)`.

We don't want to have to coordinate changes to this across runtime and framework if the sharding set changes over time.

@lewing, do you think that's possible in time for RC2?

**Review by SteveSandersonMS — APPROVED** [review]
Looks good!

Really great job on the E2E tests here. That must have been quite difficult.

File: src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs (hunk 4 above)
Comment (EgorBo) [reply src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs:90]: @pranavkm to be clear: is `kn` not expected to work in non-CJK icudt?

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts (hunk 2 above)
Comment (EgorBo) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:510]: it should be easy to add, I thought you might want to rename the files or add more variations of icudt (or user might want to do it) since it's not difficult to create an icudt for any culture(s) user wants

Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:510]: Thanks @EgorBo, that would be great! Yes it would be good if the runtime had a function to provide info about the default built-in sharding setup. Theoretically we could still vary the behavior at the application layer, but we have no plans to vary it right now.

Do you want me to file an issue on the runtime repo to represent this?

Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:510]: @pranavkm Presumably we could merge this PR as-is and then switch to the runtime logic later, just to avoid being blocked.

Comment (EgorBo) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:510]: @SteveSandersonMS I'll file a PR shortly with a prototype.

Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:510]: I see it's here: https://github.com/dotnet/runtime/pull/41795

Thanks @EgorBo!

Comment (EgorBo) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:510]: @SteveSandersonMS so should I backport that API to 5.0rc2 ?

File: src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs (hunk 4 above)
Comment (pranavkm) [reply src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs:90]: @EgorBo I would have expected it to work. This doesn't seem specific to the shards. Even if I load `icudt.dat`, there's a bunch of different cultures for which it looks like culture data is missing.

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts (hunk 6 above)
Comment (pranavkm) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:319]: > Could it ever happen because, for example, the browser says it's in a culture that .NET doesn't even know about the existence of?

From what I have tried, this is not the case. .NET creates a culture using the value specified by browser \ emscripten regardless of it's known. You get invariant assets, but do not see this error unless the culture is tweaked in .NET code.

**Comment by pranavkm** [comment]
@Pilchie, could you approve this for rc2?

**Comment by Pilchie** [comment]
Approved for .NET 5. RC2.  Are there any customers we can get to try out nightly builds once this is an installer to get some early validation and help us feel more comfortable with these changes.

**Comment by pranavkm** [comment]
@Pilchie we're going to reach out some community members on out Gitter and ask them to try out nightlies once they become available.

File: src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs (hunk 4 above)
Comment (EgorBo) [reply src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs:90]: @pranavkm sorry for the delay, the fix is here https://github.com/dotnet/icu/pull/34
I guess we need to backport it to 5.0rc2 ?

Comment (EgorBo) [reply src/Components/test/E2ETest/Tests/WebAssemblyICUShardingTest.cs:90]: the bug affects the following locales: "bn", "et", "gu", "kn", "ml", "mr", "ms", "no", "ta", "te"
